File: ui/candidate-portal/src/app/services/occupation.service.ts
/*
 * Copyright (c) 2024 Talent Catalog.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

import {Injectable} from '@angular/core';
import {environment} from '../../environments/environment';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Occupation} from '../model/occupation';
import {map} from 'rxjs/operators';

@Injectable({
  providedIn: 'root'
})
export class OccupationService {

  private apiUrl: string = environment.apiUrl + '/occupation';

  constructor(private http: HttpClient) { }

  listOccupations(): Observable<Occupation[]> {
    return this.http.get<Occupation[]>(`${this.apiUrl}`).pipe(
      map((items: Occupation[], index: number) => {
        const unknown: Occupation = items.find(x => x.id === 0);
        const i: number = items.indexOf(unknown);
        if (unknown){
          items.splice(i, 1);
          items.push(unknown);
        }
        return items;
      }
    ))
  }

}

File: server/src/main/java/org/tctalent/server/service/db/OccupationService.java
/*
 * Copyright (c) 2024 Talent Catalog.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.service.db;

import java.util.List;
import org.springframework.data.domain.Page;
import org.springframework.lang.NonNull;
import org.tctalent.server.exception.EntityExistsException;
import org.tctalent.server.exception.EntityReferencedException;
import org.tctalent.server.exception.NoSuchObjectException;
import org.tctalent.server.model.db.Occupation;
import org.tctalent.server.request.occupation.CreateOccupationRequest;
import org.tctalent.server.request.occupation.SearchOccupationRequest;
import org.tctalent.server.request.occupation.UpdateOccupationRequest;
import org.tctalent.server.util.dto.DtoBuilder;

public interface OccupationService {

    /**
     * Find occupation matching given
     * <a href="https://ilostat.ilo.org/methods/concepts-and-definitions/classification-occupation/">ISCO08 code</a>.
     * @param isco08Code ISCO code
     * @return occupation
     * @throws NoSuchObjectException if not found
     */
    @NonNull
    Occupation findByIsco08Code(String isco08Code);

    List<Occupation> listOccupations();

    Page<Occupation> searchOccupations(SearchOccupationRequest request);

    Occupation getOccupation(long id);

    Occupation createOccupation(CreateOccupationRequest request) throws EntityExistsException;

    Occupation updateOccupation(long id, UpdateOccupationRequest request) throws EntityExistsException ;

    boolean deleteOccupation(long id) throws EntityReferencedException;

    DtoBuilder selectBuilder();

}
